File: ios/TokenTracker/Notifications/SpendAlertNotifier.swift
import Foundation
import UserNotifications

/// Requests notification permission and posts the local "spend approaching
/// limit" notification. Swift sibling of the Android `SpendAlertNotifier`.
enum SpendAlertNotifier {

    /// Prompts for notification permission. Returns whether it's granted —
    /// called only when the user turns the alert on.
    static func requestAuthorization() async -> Bool {
        let center = UNUserNotificationCenter.current()
        let granted = try? await center.requestAuthorization(options: [.alert, .sound, .badge])
        return granted ?? false
    }

    /// Posts the alert immediately. No-ops if authorization was later revoked.
    static func notify(spentFormatted: String, percent: Int) async {
        let center = UNUserNotificationCenter.current()
        let settings = await center.notificationSettings()
        guard settings.authorizationStatus == .authorized ||
            settings.authorizationStatus == .provisional else { return }

        let content = UNMutableNotificationContent()
        content.title = "Spend at \(percent)% of your limit"
        content.body = "\(spentFormatted) spent so far this month."
        content.sound = .default

        let request = UNNotificationRequest(
            identifier: "spend-alert",
            content: content,
            trigger: nil // deliver now
        )
        try? await center.add(request)
    }
}

File: android/app/src/main/kotlin/studio/maximumimpact/tokencounter/notifications/SpendAlertNotifier.kt
package studio.maximumimpact.tokencounter.notifications

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import studio.maximumimpact.tokencounter.MainActivity
import studio.maximumimpact.tokencounter.R

/** Builds and posts the local "spend approaching limit" notification. */
object SpendAlertNotifier {

    const val CHANNEL_ID = "spend_alerts"
    private const val NOTIFICATION_ID = 4201
    private const val TAG = "SpendAlertNotifier"

    /** Idempotent — creates the channel if it doesn't exist (minSdk 26). */
    fun ensureChannel(context: Context) {
        val manager = context.getSystemService(NotificationManager::class.java)
        if (manager.getNotificationChannel(CHANNEL_ID) == null) {
            manager.createNotificationChannel(
                NotificationChannel(
                    CHANNEL_ID,
                    "Spend alerts",
                    NotificationManager.IMPORTANCE_DEFAULT
                ).apply {
                    description = "Alerts when your monthly spend approaches your limit."
                }
            )
        }
    }

    /**
     * Posts the alert. No-ops if POST_NOTIFICATIONS isn't granted (API 33+) —
     * the toggle requests it on opt-in, but a later revocation shouldn't crash
     * the background worker.
     */
    fun notifySpendApproaching(context: Context, spentFormatted: String, percent: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            ContextCompat.checkSelfPermission(
                context,
                Manifest.permission.POST_NOTIFICATIONS
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        ensureChannel(context)

        val openApp = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        val pendingIntent = PendingIntent.getActivity(
            context,
            0,
            openApp,
            PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_spend_alert)
            .setContentTitle("Spend at $percent% of your limit")
            .setContentText("$spentFormatted spent so far this month.")
            .setContentIntent(pendingIntent)
            .setAutoCancel(true)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .build()

        try {
            NotificationManagerCompat.from(context).notify(NOTIFICATION_ID, notification)
        } catch (e: SecurityException) {
            // Permission can be revoked between the explicit check and notify().
            Log.w(TAG, "Unable to post spend alert because notification permission was revoked.", e)
        }
    }
}
